### PR TITLE
Add Bulk Actions and Improve List/Token Generation

### PR DESCRIPTION
**Pull request by dshanske**
This improves the token list table, adding the ability to revoke tokens based on last accessed dates, as well as cleaning up expired tokens and blowing out authorization codes if there. It also adds the expiration to the list table for future expiring tokens as limited support is there.

Also, the scopes for the table generated scopes were not configurable and now they are.

### CODE REVIEW

**Comment by dshanske** [comment]
@pfefferle If you could look at this one, I'll release 3.3.2

**Comment by dshanske** [comment]
@pfefferle Added a notation re 5.2 compatibility.

File: includes/class-token-list-table.php [hunk 1]
```diff
@@ -32,37 +46,95 @@ public function prepare_items() {
...
 		}
+
 	}
 
 	public function column_default( $item, $column_name ) {
 		return $item[ $column_name ];
 	}
 
+	public function column_cb( $item ) {
+		return sprintf( '<input type="checkbox" name="tokens[]" value="%s" />', $item['token'] );
```
Comment (pfefferle) [inline includes/class-token-list-table.php]: I would recommend to output-escape the token.

File: includes/class-token-list-table.php [hunk 2]
```diff
@@ -32,37 +46,95 @@ public function prepare_items() {
...
+	}
+
 	public function process_action() {
-		$tokens = isset( $_GET['tokens'] ) ? $_GET['tokens'] : array(); // phpcs:ignore
+		$tokens = isset( $_REQUEST['tokens'] ) ? $_REQUEST['tokens'] : array(); // phpcs:ignore
```
Comment (pfefferle) [inline includes/class-token-list-table.php]: I would recommend to sanatize the request params.

File: includes/class-token-list-table.php [hunk 3]
```diff
@@ -7,15 +7,29 @@
...
 
+	public function get_bulk_actions() {
+			  return array(
+				  'revoke'   => __( 'Revoke', 'indieauth' ),
+				  'revokey'  => __( 'Revoke Tokens Last Accessed 1 Year Ago', 'indieauth' ),
+				  'revokem'  => __( 'Revoke Tokens Last Accessed 1 Month Ago', 'indieauth' ),
+				  'revokew'  => __( 'Revoke Tokens Last Accessed 1 Week Ago', 'indieauth' ),
+				  'revoked'  => __( 'Revoke Tokens Last Accessed 1 Day Ago', 'indieauth' ),
+				  'revokeh'  => __( 'Revoke Tokens Last Accessed 1 Hour Ago', 'indieauth' ),
+				  'maintain' => __( 'Clean Up Expired Tokens and Authorization Codes', 'indieauth' ),
```
Comment (pfefferle) [inline includes/class-token-list-table.php]: the keys are not very descriptive, especially "maintain".

Comment (pfefferle) [reply includes/class-token-list-table.php]: I would prever at least `revoke_*` and `cleanup` instead of `maintain`.

**Review by pfefferle — APPROVED** [review]
I have only reviewed the code, I have not tested the functionality.

File: includes/class-indieauth-token-ui.php [hunk 4]
```diff
@@ -54,7 +54,12 @@ public function new_token() {
 			exit;
 		}
 		require ABSPATH . 'wp-admin/admin-header.php';
-		$token = self::generate_local_token( $_REQUEST['client_name'] );
+		$scopes = trim( implode( ' ', $_REQUEST['scopes'] ) );
```
Comment (pfefferle) [inline includes/class-indieauth-token-ui.php]: sanatize?

**Pull request merged**